File: oxidize-python/oxidize_python/internal/gguf/writer.py
"""GGUF writer mirroring oxidize-golang/internal/gguf/writer.go."""

from __future__ import annotations

import struct
from dataclasses import dataclass
from io import BytesIO

from oxidize_python.internal.gguf.parse import align_up
from oxidize_python.internal.gguf.types import MetadataType, MetadataValue, TensorInfo


@dataclass
class WriterHeader:
    version: int = 3
    tensor_count: int = 0
    metadata_count: int = 0
    metadata: dict[str, MetadataValue] | None = None
    tensors: list[TensorInfo] | None = None
    alignment: int = 32
    data_section_start: int = 0


def encode(header: WriterHeader, body: bytes) -> bytes:
    h = header
    if h.version == 0:
        h.version = 3
    if h.alignment == 0:
        h.alignment = 32
    metadata = h.metadata or {}
    tensors = h.tensors or []
    if h.metadata_count == 0:
        h.metadata_count = len(metadata)
    if h.tensor_count == 0:
        h.tensor_count = len(tensors)

    buf = BytesIO()
    buf.write(b"GGUF")
    buf.write(struct.pack("<I", h.version))
    buf.write(struct.pack("<Q", h.tensor_count))
    buf.write(struct.pack("<Q", h.metadata_count))
    for key, value in metadata.items():
        _write_string(buf, key)
        _write_value(buf, value)
    for tensor in tensors:
        _write_string(buf, tensor.name)
        buf.write(struct.pack("<I", len(tensor.dimensions)))
        for dim in tensor.dimensions:
            buf.write(struct.pack("<Q", dim))
        buf.write(struct.pack("<I", tensor.ggml_type))
        buf.write(struct.pack("<Q", tensor.relative_offset))

    header_end = buf.tell()
    data_start = align_up(header_end, h.alignment)
    while buf.tell() < data_start:
        buf.write(b"\x00")
    if buf.tell() != data_start:
        raise ValueError("alignment mismatch")
    buf.write(body)
    return buf.getvalue()


def _write_string(buf: BytesIO, s: str) -> None:
    data = s.encode("utf-8")
    buf.write(struct.pack("<Q", len(data)))
    buf.write(data)


def _write_value(buf: BytesIO, value: MetadataValue) -> None:
    buf.write(struct.pack("<I", int(value.type)))
    _write_value_body(buf, value)


def _write_value_body(buf: BytesIO, value: MetadataValue) -> None:
    match value.type:
        case MetadataType.UINT8:
            buf.write(struct.pack("<B", value.uint64 & 0xFF))
        case MetadataType.INT8:
            buf.write(struct.pack("<b", value.int64))
        case MetadataType.UINT16:
            buf.write(struct.pack("<H", value.uint64 & 0xFFFF))
        case MetadataType.INT16:
            buf.write(struct.pack("<h", value.int64))
        case MetadataType.UINT32:
            buf.write(struct.pack("<I", value.uint64 & 0xFFFFFFFF))
        case MetadataType.INT32:
            buf.write(struct.pack("<i", value.int64))
        case MetadataType.FLOAT32:
            buf.write(struct.pack("<f", value.float64))
        case MetadataType.BOOL:
            buf.write(struct.pack("<B", 1 if value.bool else 0))
        case MetadataType.STRING:
            _write_string(buf, value.string)
        case MetadataType.ARRAY:
            if not value.array:
                buf.write(struct.pack("<I", int(MetadataType.UINT8)))
                buf.write(struct.pack("<Q", 0))
            else:
                buf.write(struct.pack("<I", int(value.array[0].type)))
                buf.write(struct.pack("<Q", len(value.array)))
                for element in value.array:
                    _write_value_body(buf, element)
        case MetadataType.UINT64:
            buf.write(struct.pack("<Q", value.uint64))
        case MetadataType.INT64:
            buf.write(struct.pack("<q", value.int64))
        case MetadataType.FLOAT64:
            buf.write(struct.pack("<d", value.float64))
        case _:
            raise ValueError(f"unsupported metadata type {value.type}")

File: oxidize-golang/internal/gguf/writer.go
package gguf

import (
	"bytes"
	"encoding/binary"
	"fmt"
)

// WriterHeader describes a complete GGUF file to be encoded.
type WriterHeader struct {
	Version          uint32
	TensorCount      uint64
	MetadataCount    uint64
	Metadata         map[string]MetadataValue
	Tensors          []TensorInfo
	Alignment        uint64
	DataSectionStart uint64
}

// Encode serializes the given GGUF header and returns the final file bytes
// consisting of the header followed by `body` (which is assumed to be already
// aligned to DataSectionStart).
func Encode(h WriterHeader, body []byte) ([]byte, error) {
	if h.Version == 0 {
		h.Version = 3
	}
	if h.Alignment == 0 {
		h.Alignment = 32
	}
	if h.MetadataCount == 0 {
		h.MetadataCount = uint64(len(h.Metadata))
	}
	if h.TensorCount == 0 {
		h.TensorCount = uint64(len(h.Tensors))
	}
	var buf bytes.Buffer
	// Magic
	buf.WriteString("GGUF")
	// Version
	if err := binary.Write(&buf, binary.LittleEndian, h.Version); err != nil {
		return nil, err
	}
	// Tensor count
	if err := binary.Write(&buf, binary.LittleEndian, h.TensorCount); err != nil {
		return nil, err
	}
	// Metadata count
	if err := binary.Write(&buf, binary.LittleEndian, h.MetadataCount); err != nil {
		return nil, err
	}
	// Metadata
	for k, v := range h.Metadata {
		if err := writeString(&buf, k); err != nil {
			return nil, err
		}
		if err := writeValue(&buf, v); err != nil {
			return nil, err
		}
	}
	// Tensor infos
	for _, t := range h.Tensors {
		if err := writeString(&buf, t.Name); err != nil {
			return nil, err
		}
		if err := binary.Write(&buf, binary.LittleEndian, uint32(len(t.Dimensions))); err != nil {
			return nil, err
		}
		for _, d := range t.Dimensions {
			if err := binary.Write(&buf, binary.LittleEndian, d); err != nil {
				return nil, err
			}
		}
		if err := binary.Write(&buf, binary.LittleEndian, t.GGMLType); err != nil {
			return nil, err
		}
		if err := binary.Write(&buf, binary.LittleEndian, t.RelativeOffset); err != nil {
			return nil, err
		}
	}
	// Pad to data section start
	headerEnd := uint64(buf.Len())
	dataStart, err := alignUp(headerEnd, h.Alignment)
	if err != nil {
		return nil, err
	}
	for uint64(buf.Len()) < dataStart {
		buf.WriteByte(0)
	}
	if uint64(buf.Len()) != dataStart {
		return nil, fmt.Errorf("alignment mismatch")
	}
	// Tensor relative offsets must be relative to data section start
	for i := range h.Tensors {
		h.Tensors[i].RelativeOffset = 0
	}
	buf.Write(body)
	return buf.Bytes(), nil
}

func writeString(buf *bytes.Buffer, s string) error {
	if err := binary.Write(buf, binary.LittleEndian, uint64(len(s))); err != nil {
		return err
	}
	_, err := buf.WriteString(s)
	return err
}

func writeValue(buf *bytes.Buffer, v MetadataValue) error {
	if err := binary.Write(buf, binary.LittleEndian, uint32(v.Type)); err != nil {
		return err
	}
	return writeValueBody(buf, v)
}

func writeValueBody(buf *bytes.Buffer, v MetadataValue) error {
	switch v.Type {
	case MetadataUint8:
		return binary.Write(buf, binary.LittleEndian, uint8(v.Uint64))
	case MetadataInt8:
		return binary.Write(buf, binary.LittleEndian, int8(v.Int64))
	case MetadataUint16:
		return binary.Write(buf, binary.LittleEndian, uint16(v.Uint64))
	case MetadataInt16:
		return binary.Write(buf, binary.LittleEndian, int16(v.Int64))
	case MetadataUint32:
		return binary.Write(buf, binary.LittleEndian, uint32(v.Uint64))
	case MetadataInt32:
		return binary.Write(buf, binary.LittleEndian, int32(v.Int64))
	case MetadataFloat32:
		return binary.Write(buf, binary.LittleEndian, float32(v.Float64))
	case MetadataBool:
		var b uint8
		if v.Bool {
			b = 1
		}
		return binary.Write(buf, binary.LittleEndian, b)
	case MetadataString:
		return writeString(buf, v.String)
	case MetadataArray:
		if len(v.Array) == 0 {
			// write dummy element type
			if err := binary.Write(buf, binary.LittleEndian, uint32(MetadataUint8)); err != nil {
				return err
			}
			return binary.Write(buf, binary.LittleEndian, uint64(0))
		}
		et := uint32(v.Array[0].Type)
		if err := binary.Write(buf, binary.LittleEndian, et); err != nil {
			return err
		}
		if err := binary.Write(buf, binary.LittleEndian, uint64(len(v.Array))); err != nil {
			return err
		}
		for _, e := range v.Array {
			if err := writeValueBody(buf, e); err != nil {
				return err
			}
		}
		return nil
	case MetadataUint64:
		return binary.Write(buf, binary.LittleEndian, v.Uint64)
	case MetadataInt64:
		return binary.Write(buf, binary.LittleEndian, v.Int64)
	case MetadataFloat64:
		return binary.Write(buf, binary.LittleEndian, v.Float64)
	default:
		return fmt.Errorf("unsupported metadata type %v", v.Type)
	}
}

// alignUp is provided by parse.go
